### PR TITLE
Keep more type information in Lambda

### PR DESCRIPTION
**Pull request by alainfrisch**
This PR adds more propagation of "value kinds" (used to drive float unboxing, in particular) through the compiler:

  - This is based on a version of #1029 from @chambart merged with trunk, i.e. keeping value kinds on lambda functions (parameters and return values).

  - Adding value kinds on parameters of staticcatch.

  - ~~Unboxing across staticcatch handlers, as in [MPR#7017](https://caml.inria.fr/mantis/view.php?id=7017), but based on value kinds from the item above.  (For now, float only, not boxed numbers.)~~

The first point is somehow independent from the other ones, but (1) it goes in the same direction as the second point; and (2) all together they will allow to make sure that functions compiled "statically" as in #2143 will not require boxing when jumping to a shared continuation.

Variants of #1322 (unboxed calling conventions) and other optimizations could also benefit from keeping more value kinds in lambda/clambda.

**The text below does not apply any more to this PR**

In its current state, this PR already avoids boxing `b` in;

````ocaml
let f z x y =
  let a, b =
    if z then (x * 2, y *. 3.)
    else (x, y)
  in
  float a -. b
````

The tuple binding is compiled to a staticcatch, and this optimization now makes sure to annotate the parameters of the staticcatch with value kinds derived from the pattern type.  (This only works without flambda.  Extra work will be needed to propagate -- and hopefully enrich -- the value kind annotations through flambda.)

Similarly, the PR avoids some boxing with or-pattern, such as for `x` in:

````ocaml
type t =
  {
    x: float;
    y: float;
  }

let f = function
  | Some {x; y = 0.} | Some {x = 0.; y = x} -> x +. x
  | _ -> 0.
````

The unboxing decision for staticcatch parameters (in Cmmgen) is currently only based on value kind annotations (inherit from Lambda).  One bad case is when all staticraise sites naturally produce a boxed version (e.g. by calling a float-returning function, or reading from a boxed data structure), and the statichandler can use this boxed version directly.  With the new heuristics, those floats will be unboxed and then reboxed when accessed in the handler.  This is similar to what happens for float mutable variables, which are always unboxed (independently of the bound expression), and I think one can live with that.~~

### CODE REVIEW

**Comment by alainfrisch** [comment]
I've edited the description and removed the [WIP] marker.  This is ready for review.

**Comment by alainfrisch** [comment]
(Note: CI reported on #1029 a problem while compiling Camlp4, because of "empty pattern matching", not allowed by the normal syntax.  I've added a commit which should fix this problem.  AppVeyor/Travis don't compile Camlp4 anymore.)

**Comment by alainfrisch** [comment]
Commit  998d831 avoids boxing of `x` in:

````ocaml
type t =
  {
    x: float;
    y: float;
  }

let f = function
  | Some {x; y = 0.} | Some {x = 0.; y = x} -> x +. x
  | _ -> 0.
````

**Comment by alainfrisch** [comment]
Would someone be able to review this PR (perhaps @chambart , @lpw25 or @mshinwell)?

File: asmcomp/un_anf.ml [hunk 1]
```diff
@@ -156,7 +156,7 @@ let make_var_info (clam : Clambda.ulambda) : var_info =
       List.iter loop args
     | Ucatch (static_exn, vars, body, handler) ->
       ignore_int static_exn;
-      ignore_var_with_provenance_list vars;
+      ignore_var_with_provenance_list (List.map fst vars);
```
Comment (chambart) [inline asmcomp/un_anf.ml]: That function does nothing but asserts the type. you should change the definition instead:
```
let ignore_var_with_provenance_list (_ : (VP.t  * value_kind) list) = ()
```

File: bytecomp/lambda.ml [hunk 2]
```diff
@@ -788,10 +793,13 @@ let rec map f lam =
...
 
-let bind str var exp body =
+let bind_with_value_kind str (var, kind) exp body =
   match exp with
     Lvar var' when Ident.same var var' -> body
-  | _ -> Llet(str, Pgenval, var, exp, body)
+  | _ -> Llet(str, kind, var, exp, body)
+
+let bind str var exp body =
+  bind_with_value_kind str (var, Pgenval) exp body
```
Comment (chambart) [inline bytecomp/lambda.ml:802]: I've looked at the use points for this function. They are all in `Matching`. Most of them do have kind information. I think it would probably be better to update the uses rather than the definition.

File: asmcomp/un_anf.ml (hunk 1 above)
Comment (alainfrisch) [reply asmcomp/un_anf.ml]: The function was used also for functions parameters.  I've fixed that by keeping  value kinds on those parameters in Clambda as well :-) (and also for the function return).

File: bytecomp/lambda.ml (hunk 2 above)
Comment (alainfrisch) [reply bytecomp/lambda.ml:802]: I've switched two call sites to compute the value_kind, but all others in Matching, there does not seem to be a trivial way to compute the value_kind without more refactoring.

File: bytecomp/translcore.ml [hunk 3]
```diff
@@ -899,7 +904,7 @@ and transl_match e arg pat_expr_list partial =
     Lstaticcatch
       (Ltrywith (Lstaticraise (static_exception_id, body), id,
                  Matching.for_trywith (Lvar id) exn_cases),
-       (static_exception_id, val_ids),
+       (static_exception_id, (List.map (fun id -> id, Pgenval) val_ids)),
```
Comment (chambart) [inline bytecomp/translcore.ml]: The kind information is available for that case:
```OCaml
let f x =
  match x () with
  | exception Not_found -> 1.
  | n -> n
```
The catch should be annotated with float here
```
       (function x/5 : float
         (catch
           (try (exit 1 (apply x/5 0a)) with exn/9
             (if (== exn/9 (field 8 (global Stdlib!))) 1. (reraise exn/9)))
          with (1 n/6) n/6)))
```

File: asmcomp/cmmgen.ml [hunk 4]
```diff
@@ -60,6 +65,15 @@ let add_unboxed_id id unboxed_id bn env =
...
   }
 
+let add_catch_types i typs env =
+  { env with
+    catch_types = IntMap.add i typs env.catch_types
+  }
+
+let catch_types i env =
+  try IntMap.find i env.catch_types
+  with Not_found -> []
```
Comment (chambart) [inline asmcomp/cmmgen.ml]: That should probably be a fatal_error here, rather than in the assertion bellow

File: asmcomp/cmmgen.ml [hunk 5]
```diff
@@ -2030,6 +2052,33 @@ let rec transl env e =
...
+      (fun (id, k) (env, typs, ids) ->
+         match k with
+         | Pfloatval ->
+             let dbg = Debuginfo.none in
+             let unboxed_id = V.create_local (VP.name id) in
+             let typ = Boxed_float dbg in
+             add_unboxed_id (VP.var id) unboxed_id typ env,
+             Some typ :: typs,
+             (* keep provenance from id? *)
+             (VP.create unboxed_id, Cmm.typ_float) :: ids
```
Comment (chambart) [inline asmcomp/cmmgen.ml]: That unboxing might be too aggressive. it doesn't take the use into account for the decision.

File: asmcomp/cmmgen.ml [hunk 6]
```diff
@@ -2030,6 +2052,33 @@ let rec transl env e =
...
+    List.fold_right
+      (fun (id, k) (env, typs, ids) ->
+         match k with
+         | Pfloatval ->
+             let dbg = Debuginfo.none in
+             let unboxed_id = V.create_local (VP.name id) in
+             let typ = Boxed_float dbg in
+             add_unboxed_id (VP.var id) unboxed_id typ env,
+             Some typ :: typs,
+             (* keep provenance from id? *)
```
Comment (chambart) [inline asmcomp/cmmgen.ml]: The current situation for the other unboxing cases is not to propagate, but the right choice is probably to propagate here and there. @mshinwell can you confirm ?

**Comment by chambart** [comment]
I would have separated the propagation/unboxing in 2 PR, but it doesn't really matter.
I had a few remarks, but otherwise I have no objection to the propagation part and the code is ok.
The unboxing part, I think, could have been handled exactly like the non-mutable lets.

File: asmcomp/cmmgen.ml (hunk 6 above)
Comment (mshinwell) [reply asmcomp/cmmgen.ml]: Yes, I think the provenance from `id` should be preserved here.

**Comment by mshinwell** [comment]
Could we please separate the unboxing part from the propagation of kinds in this patch?  They seem like distinct features.

File: asmcomp/cmmgen.ml (hunk 4 above)
Comment (alainfrisch) [reply asmcomp/cmmgen.ml]: There are a number of `ccatch` created elsewhere (e.g. for `Ufor`), and they all have an empty list of arguments.

Comment (alainfrisch) [reply asmcomp/cmmgen.ml]: Anyway, this will be for another PR!

**Comment by alainfrisch** [comment]
Ok, let's try to get the (tedious but mechanical) propagation stuff first.  I've reverted the commit about the actual unboxing and will submit another PR later.

File: bytecomp/translcore.ml (hunk 3 above)
Comment (alainfrisch) [reply bytecomp/translcore.ml]: Thanks, done.

**Comment by alainfrisch** [comment]
All comments addressed.  I've reviewed the part originally written by @chambart , and @chambart looked at the part I added.  @chambart : if you are happy with the current state, can you add a formal review?

File: asmcomp/cmmgen.ml (hunk 4 above)
Comment (chambart) [reply asmcomp/cmmgen.ml]: You don't have clambda version for those catches, so catch_types never get called for them.

File: bytecomp/translcore.ml [hunk 7]
```diff
@@ -924,17 +924,25 @@ and transl_match e arg pat_expr_list partial =
       assert (static_handlers = []);
       Matching.for_multiple_match e.exp_loc (transl_list argl) val_cases partial
     | {exp_desc = Texp_tuple argl}, _ :: _ ->
-      let val_ids = List.map (fun _ -> Typecore.name_pattern "val" []) argl in
-      let lvars = List.map (fun id -> Lvar id) val_ids in
-      static_catch (transl_list argl) val_ids
-        (Matching.for_multiple_match e.exp_loc lvars val_cases partial)
+        let val_ids =
```
Comment (chambart) [inline bytecomp/translcore.ml:927]: Why did you change the indentation depth here ? I know that 4 is the default, but the rest of the code is 2 here

Comment (alainfrisch) [reply bytecomp/translcore.ml:927]: I didn't; emacs+ocp-indent did, and I'm a bit too lazy to find how to disable this behavior.  You can ask GitHub to hide whitespace changes if this helps reviewing.

File: asmcomp/flambda_to_clambda.ml [hunk 8]
```diff
@@ -527,7 +527,9 @@ and to_clambda_set_of_closures t env
     in
     { label = Compilenv.function_label closure_id;
       arity = Flambda_utils.function_arity function_decl;
-      params = List.map (fun var -> VP.create var) (params @ [env_var]);
+      params =
+        List.map (fun var -> VP.create var, Lambda.Pgenval) (params @ [env_var]);
```
Comment (chambart) [inline asmcomp/flambda_to_clambda.ml]: check_typo is complaining about this line (81 characters)

**Comment by chambart** [comment]
I think it's good to go when the CI is happy (the only failure is check_typo).

**Comment by shindere** [comment]
This breaks the `tests/translprim/array_spec.ml` test when the
compiler is configured with -no-flat-float-array.

More precisely: there is a difference between the expected and the obtained
output:

-      (array.length[addr] addr_a) (function a (array.length[addr] a))
+      (array.length[addr] addr_a) (function a : int (array.length[addr] a))



Shall I just go ahead and update the reference file or is there something
deeper to do here?

**Comment by alainfrisch** [comment]
No, nothing deep, this simply needs to be updated!

**Comment by shindere** [comment]
Alain Frisch (2018/11/23 07:23 -0800):
> No, nothing deep, this simply needs to be updated!

Done in commit eb951f1a6e9f78f9249336c8edee2f1a537799da

**Comment by alainfrisch** [comment]
Thanks! @shindere!

**Comment by shindere** [comment]
Alain Frisch (2018/11/23 08:32 -0800):
> Thanks! @shindere!

No problem :)

**Comment by buergerling** [comment]
I'd like to use the -dtypedtree flag and a modified printtyped.ml to transpile from ocaml into another language without such advanced pattern matching capabilities.

Is there a pattern match compiler keeping all type information?

**Comment by gasche** [comment]
@buergerling I would recommend instead that you link to `compilerlibs` and implement your own driver program; `driver/compile_common.mli` shows you how to parse and type-check a module, and then you can use your own code to turn the typedtree into an AST for your output language -- or print it directly. (The Compierlibs API is not stable, so this couples you to a specific compiler version, but then so does forking the compiler as you have in mind.)

The compilation of pattern-matching is done during the Typedtree->(untyped)Lambda translation, so it does not preserve type information. This means that you would probably have to rewrite your own pattern-matching compiler yourself. You could also fork OCaml's pattern-matching compiler to emit a typed representation, and I think it would be doable, but it's a complex piece of the codebase.
